### PR TITLE
seaborn: complete and fix axisgrid module

### PR DESCRIPTION
**Pull request by hamdanal**

### CODE REVIEW

File: stubs/seaborn/seaborn/axisgrid.pyi [hunk 1]
```diff
@@ -221,22 +378,22 @@ def pairplot(
...
+    data: DataSource | _DataSourceWideForm | None = None,
     *,
-    x: Incomplete | None = None,
-    y: Incomplete | None = None,
-    hue: Incomplete | None = None,
-    kind: str = "scatter",  # ideally Literal["scatter", "kde", "hist", "hex", "reg", "resid"] but it is checked with startswith
+    x: ColumnName | _Vector | None = None,
+    y: ColumnName | _Vector | None = None,
+    hue: ColumnName | _Vector | None = None,
+    kind: Literal["scatter", "kde", "hist", "hex", "reg", "resid"] = "scatter",
```
Comment (hamdanal) [inline stubs/seaborn/seaborn/axisgrid.pyi:386]: This was changed to `str` earlier thanks to the [feedback](https://github.com/python/typeshed/pull/10721#discussion_r1355999012) on the PR that introduced seaborn stubs. Upon further investigation, the function seems to [only accepts this set of strings](https://github.com/mwaskom/seaborn/blob/908ca95137c0e73bb6ac9ce9a8051577b6453138/seaborn/axisgrid.py#L2223) (also verified in the repl).

File: stubs/seaborn/seaborn/axisgrid.pyi [hunk 2]
```diff
@@ -127,15 +284,15 @@ class PairGrid(Grid):
     axes: NDArray[Incomplete]  # two-dimensional array of `Axes`
     data: DataFrame
     diag_sharey: bool
-    diag_vars: NDArray[Incomplete] | None  # array of `str`
-    diag_axes: NDArray[Incomplete] | None  # array of `Axes`
+    diag_vars: list[str] | None
+    diag_axes: list[Axes] | None
```
Comment (hamdanal) [inline stubs/seaborn/seaborn/axisgrid.pyi:288]: These are strings, not sure why I typed them as arrays earlier

File: stubs/seaborn/seaborn/axisgrid.pyi [hunk 3]
```diff
@@ -60,7 +211,7 @@ class FacetGrid(Grid):
     hue_kws: dict[str, Any]
     def __init__(
         self,
-        data: DataFrame,
+        data: DataFrame | _SupportsDataFrame,
```
Comment (hamdanal) [inline stubs/seaborn/seaborn/axisgrid.pyi:214]: seaborn version 0.13 started supporting other data frames that support the `__dataframe__` protocol: https://seaborn.pydata.org/whatsnew/v0.13.0.html#support-for-alternate-dataframe-libraries

**Review by srittau — APPROVED** [review]
I did not compare this to the implementation, but all the changes LGTM.

**Pull request merged**